### ISSUE TITLE
Ability to add PIN/password when unsealing with TPM

### ISSUE THREAD

**Issue by wiktor-k** [electrickite/luks-tpm2#4]
Hello,

I'd like to ask if it would be possible to use additional PIN/password when unlocking the keyfile with TPM?

The PIN wouldn't need to be long or complex as the TPM provides brute-force attack resistance. Something similar can be seen [in Bitlocker](https://www.howtogeek.com/262720/how-to-enable-a-pre-boot-bitlocker-pin-on-windows/).

**Comment by electrickite** [comment]
@wiktor-k How would you see this working? The TPM 2.0 spec is much more flexible than the 1.1, meaning there are multiple ways this could be accomplished. 

Currently, I don’t see a way to do this with the 3.X stable version of tpm2-tools. This would likely need to be some kind of combined PCR and password auth policy but, AFAIK, there is no way to create one without using a non-kernel resource manager.

That said, I am by no means a TPM expert, so if you can think of a way to add password auth in addition to PCR auth, I would be happy to build it in to luks-tpm2!

**Comment by wiktor-k** [comment]
Hmm... to be honest I'm not entirely sure how to approach it myself but I've noticed a lot of tpm2 commands (e.g. `tpm2_create` and `tpm2_createprimary`) take `-pwdk` parameter that is described as `The password for key, optional.` (`man tpm2_createprimary` shows more info)

I'm thinking of creating primary key protected with password, then sealing and unsealing some random key to see if tpm2 asks for the key password but I need to glue together what is in luks-tpm2 and mkinitcpio hook so it may take some time (I'm not a TPM expert myself too :) ).

**Comment by electrickite** [comment]
Did some more tinkering with this recently. I think we can make this work for keys stored on disk. There is no way I can currently find to protect an NVRAM index (or persisted object) with both a PCR policy AND a password. For keys stored on disk, like you suggest we can place a password on the parent key and use that as a preboot PIN.

This will also need changes to the boot process. On Arch a new version of `mkinitcpio-tpm2-encrypt`

Unfortunately, right now there is a kernel bug with the Linux 5.0 TPM driver that is preventing me from making any progress on this. More to come once that is resolved...

**Comment by wiktor-k** [comment]
Thanks for the update @electrickite!

> Unfortunately, right now there is a kernel bug with the Linux 5.0 TPM driver that is preventing me from making any progress on this. More to come once that is resolved...

Oh. So that's why the unlock stopped working after upgrade to 5.0.

I've recently seen an article about [Extracting BitLocker keys from a TPM](https://pulsesecurity.co.nz/articles/TPM-sniffing) that describes a technique to get the TPM decrypted drive keys. One of the countermeasures described there was additional PIN:

> Enabling BitLocker with a TPM+PIN protector should mitigate this vulnerability, however user’s will be required to enter a PIN at boot.

Have a nice day! :wave:

**Comment by electrickite** [comment]
> Extracting BitLocker keys from a TPM

I saw that as well. Thought immediately of this issue :)

**Comment by electrickite** [comment]
Version 1.1.1 adds the ability to specify the password for the parent key. While not ideal, this does meet the basic need. Without the parent password, the sealed file cannot be loaded.

`mkinitcpio-tpm2-encrypt` has also been updated with an additional kernel parameter, `tpmprompt=1` that will prompt for the parent key password during boot.

If you try this solution, be sure to have luks-tpm2 1.1.1, mkinitcpio-tpm2-encrypt 1.3.0 and kernel 5.0.10.

**Comment by wiktor-k** [comment]
I just tested that and it works~ :rocket: (the setup took a little bit longer than I anticipated as there are 2 packages and 2 places to configure + the actual TPM commands to run but now I think it'll run well without supervision).

There is a minor thing, `luks-tpm2 init` asks for `[Y/n]` but doesn't treat enter (empty response) as `Y` and it's not clear it didn't do anything.

I've got some other minor things, if you don't mind I can send you them via e-mail or so...

Thanks for your hard work, I really appreciate it! :wave:
